libvisual-plugins/plugins/input/alsa/input_alsa.c

```diff
@@ -221,14 +222,19 @@ int inp_alsa_upload (VisPluginData *plugin, VisAudio *audio)
 {
 	int16_t data[PCM_BUF_SIZE];
 	alsaPrivate *priv = NULL;
-	int rcnt;
+	const snd_pcm_uframes_t frames_wanted = sizeof(data) / sizeof(data[0]) / inp_alsa_var_channels;
```

Comment:
This is another integer division. May be better to use sizeof(data) / (sizeof(data[0]) * inp_alsa_var_channels).

Reply:
@kaixiong we are reversing the two multiplications contained in `int16_t data[PCM_BUF_SIZE]` here and then dividing by the number of channels, literally `(2 * 4096) // 2 // 2` i.e. `(2 * (2 * 2048)) // 2 // 2` i.e. `2048`. As long as `PCM_BUF_SIZE` is an even number, there is no chance for loss here. If it ever is change to something odd, then we lose one sample with or without the braces. Hence my vote for keeping as is, what do you think?

Reply:
@hartwork, sorry I was a bit too rushed with my original comment. We only support 2 channels at most at the moment, and with that int16_t, we should have divisibility (no loss) in every intermediate step.

Reply:
@kaixiong PS: I can still make that^^ adjustment, if you prefer, no hard feelings about it.

Reply:
@kaixiong only saw your latest comment now. So which version do we use?

Reply:
@hartwork, sorry for the confusion. I was happily adding the in-code comments without realising that it didn't need a final submit to show! I think it's okay as is, but if we want to make it a _teeney weeney_ bit nicer, there's a macro called VISUAL_TABLESIZE() for the first part of the equation:

```
const snd_pcm_uframes_t = VISUAL_TABLE_SIZE(data) / inp_alsa_var_channels;
```

I leave it up to you to decide :)

Reply:
@kaixiong, I found `VISUAL_TABLESIZE` now but the "table" part of the name seems more misleading than helpful, since there is not "table" here. My vote for making a copy (we cannot rename in 0.4.x) named `VISUAL_ARRAY_SIZE` for universal use that is more self-explaining, but outside of this pull request fist. I'll leave that part as is for now then.

Reply:
> My vote for making a copy (we cannot rename in 0.4.x) named `VISUAL_ARRAY_SIZE` for universal use that is more self-explaining, but outside of this pull request fist. I'll leave that part as is for now then.
@hartwork, agreed.